Add parallel test execution

bitwiseman opened the pull request:
Speed up build by executing in parallel on multiple nodes.

bitwiseman commented:
I don't understand what failed here.

bitwiseman commented:
@MarkEWaite - Thoughts?

MarkEWaite commented:
When I tried to run that in my Jenkins test environment (intentional mix of CentOS, Debian, Ubuntu, and Windows), it failed with the message:

> java.io.IOException: Cannot run program "nohup" (in directory "C:\J\workspace\git-plugin-multi-pipeline\master-PR411-parallel-tests"): CreateProcess error=2, The system cannot find the file specified

I don't see any reference to "nohup" in the Jenkinsfile, so I assume there is some Linux dependency somewhere else in that process.

I'll try some further experiments and let you know what I see.

Did you see any improvement in the time from job start to job finish with this change?

bitwiseman commented:
Just to double check, do you have the parallell test executor plugin installed?  
( https://wiki.jenkins-ci.org/display/JENKINS/Parallel+Test+Executor+Plugin)  

I saw roughly a 50% reduction in total time from start to finish.

MarkEWaite commented:
I saw an improvement the one time the tests passed.  Unfortunately I also saw cases where the wrong jdk was selected by the parallel jobs. I will need to investigate further 

I also saw cases where the parallel test jobs failed to install the mvn tool.  I'll need to investigate that further as well

bitwiseman commented:
I saw none of those issues... What version of jenkins and plugins are you running?

MarkEWaite commented:
I'm running Jenkins 1.651.2 with most recent releases of all the plugins.  I also run tests in a Jenkins 1.651.2 Docker instance so that I have a shareable example of the configuration.

I'm intentionally running Windows and Linux, and intentionally running slaves with various versions of git (CentOS 6, CentOS 7, Debian 7, Debian 8, Ubuntu 14.04, Windows 10).  Once I know more about the environment where the failures happen, I'll share it with you here.

bitwiseman commented:
Okay, I'm running jenkins 2.6.  And I'm not using docker this time.

I look forward to hearing what you find out.

rtyler commented on Jenkinsfile:
```diff
@@ -10,19 +10,56 @@ node {
 
+void runParallelTests() {
+  /* Request the test groupings.  Based on previous test exection. */
+  def splits = splitTests parallelism: [$class: 'CountDrivenParallelism', size: 4], generateInclusions: true
```
Why is the split size four? It seems arbitrary, which is probably fine, but should be documented as such

rtyler commented on Jenkinsfile:
```diff
@@ -10,19 +10,56 @@ node {
+      node {
+        checkout scm
+
+        sh 'mvn clean -B -V -U -e'
```
A `clean` was implemented above already, I don't understand why this is here.

Also, why doesn't this method use the `mvn()` helper method that was defined already in this `Jenkinsfile` to ensure that the right tool installer/JDK version is used?

Otherwise, I think this would be relying on the system JDK

MarkEWaite replied:
Thanks for detecting that!  I didn't understand why the "mvn" tool was not being auto-installed on the machines that were running tests.  I think your comment may explain that observed behavior, since the function was not called to compute the mvn tool on that node.

bitwiseman replied:
doh.  C&P error. Fixing.

rtyler commented on Jenkinsfile:
```diff
@@ -10,19 +10,56 @@ node {
+        def mavenInstall = 'mvn install -B -V -U -e -Dsurefire.useFile=false -Dmaven.test.failure.ignore=true'
+
+        /* Write include or exclude file for maven tests.  Contents provided by splitTests. */
+        if (split.includes) {
```
Can you link to the documentation or something which expands more on what these includes/excludes are and how they would have been defined?

bitwiseman replied:
There's a demo file, but the readme and other docs are a bit sparse.  Filed https://issues.jenkins-ci.org/browse/JENKINS-35682.  In the meanwhile, I'll have this point to the main plugin page and demo.

rtyler commented on Jenkinsfile:
```diff
@@ -10,19 +10,56 @@ node {
+  for (int i = 0; i < splits.size(); i++) {
+    def split = splits[i]
+
+    testGroups["split${i}"] = {
```
IMO ahead of this declaration should be some commentary explaining what it is we're doing here and why

bitwiseman replied:
Sure, but it's going to get a bit involved.

bitwiseman commented:
Updated with mvn fix.   Still need last comments.

bitwiseman commented:
Commenting done, failure doesn't seem to have anything to do with this change.

rtyler commented:
@bitwiseman will you file an INFRA ticket for the new plugin requirements this will introduce to ci.jenkins.io?

bitwiseman commented:
https://issues.jenkins-ci.org/browse/INFRA-801

bitwiseman commented:
@MarkEWaite - Can we merge this?

MarkEWaite commented:
@bitwiseman not until I understand why the results in my test and development environment are so widely variable.  I've not yet seen consistent cases where the parallel test execution passes and runs any faster than serial test execution, even with multiple machines contributing to execution.

This weekend will be focused on other areas of the git plugin, so it will be at least a week before I revisit this pull request.

bitwiseman commented:
Ah, I thought that the "tool" fix addressed the instabilities.

bitwiseman commented:
Added timeouts.

bitwiseman commented:
@MarkEWaite - 
I understand this is lower priority than other items on your radar.  
Is there a platform or two that would be most useful for me to try this on to get you more information? 

As a side note:
The checked in versions of this pipeline are actually live at https://ci.jenkins.io/job/Plugins/job/git-plugin/ .  @rtyler and I have been working on this to help define to a more systematic plugin build pipeline.  The importance and non-trivial build/testing requirements of the git plugin make it an ideal place to develop broader plugin CI/CD support in ci.jenkins.io.  In short order, this will also help make your life as maintainer easier.

MarkEWaite commented:
The most interesting platforms (for me) are
- Ubuntu 14.04 with the PPA which installs the latest git build when released (currently 2.9.0)
- CentOS 6.7 with the default git install (because it is so old, yet used by many Jenkins users)
- Windows with the latest git build (currently 2.9.0) with a JNLP client running on the desktop
- Windows with an old git build (1.9.4 ish) with a JNLP client running as a service

Those are configurations I test from my home environment (along with two Debian variants and Red Hat 7.2).

Configurations I'd like to test but can't due to lack of hardware include:
- MacOS with the standard git version in their bundle
- FreeBSD with the standard git version in their bundle
- Solaris with the standard git version in their bundle
- AIX with the standard git version in their bundle
- HP-UX with the standard git version in their bundle

bitwiseman commented:
I'll try testing this locally in a VM on the first two, ubuntu and CentOS, and get back to you.

MarkEWaite commented:
I'd also be willing to merge it to the 3.0.0-beta branch if that makes it any easier for you to test it.

bitwiseman commented:
@MarkEWaite - Merging to 3.0.0-beta would be great.  That will get it picked up by ci.jenkins.io and get us more data points.

MarkEWaite commented:
Pushed to the 3.0.0-beta branch (after rebasing, so that I can clearly see what I need to revert if there are problems with it).

MarkEWaite commented:
@bitwiseman failures are visible on the [3.0.0-beta job build 11](https://ci.jenkins.io/job/Plugins/job/git-plugin/job/3.0.0-beta/11/console) and [3.0.0-beta job build 12](https://ci.jenkins.io/job/Plugins/job/git-plugin/job/3.0.0-beta/12/console).  Those failures match the failures I see frequently.

I'm not going to investigate those failures.  I'll continue my focus on the other pull requests and assume you will either investigate them to understand the root cause, or close this pull request.  If you close this pull request, that will be my sign to revert that change from the 3.0.0-beta branch.

I don't feel any compelling push to revert the parallel test execution from the 3.0.0-beta branch so long as you're continuing to investigate the issues.

bitwiseman commented:
@MarkEWaite - Thanks for the heads up.

bitwiseman commented:
@MarkEWaite, @rtyler - I asked for help on IRC, it looks like an out of memory error:

```
Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000007e8d80000, 320864256, 0) failed; error='Cannot allocate memory' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (malloc) failed to allocate 320864256 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /home/jenkins/agent-workspace/workspace/Plugins/git-plugin/3.0.0-beta@4/hs_err_pid12574.log
Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000007e8c80000, 319815680, 0) failed; error='Cannot allocate memory' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (malloc) failed to allocate 319815680 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /home/jenkins/agent-workspace/workspace/Plugins/git-plugin/3.0.0-beta@5/hs_err_pid12729.log
Tests run: 696, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13.741 sec - in hudson.plugins.git.GitChangeSetPluginHistoryTest
Running hudson.plugins.git.GitStatusTest
Killed
```

There's at least three things we can try: 
a. Add memory (or more swap space) to VM's
b. Switch to jdk8 or investigate java settings
c. Analyze tests to see what is pulling such a large chunk of memory. 

I think (a) is the most reasonable place to start.  It makes sense that a machine would run short on memory when it runs multiple processes, and it is relatively easy to try.

MarkEWaite commented:
I think c is the most likely to be fruitful.  Our of memory had been a
common occurrence since the upgrade to the 2.x pom

On Wed, Jul 6, 2016, 1:12 PM Liam Newman <notifications@github.com> wrote:

> @MarkEWaite <https://github.com/MarkEWaite>, @rtyler
> <https://github.com/rtyler> - I asked for help on IRC, it looks like an
> out of memory error:
>
> Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000007e8d80000, 320864256, 0) failed; error='Cannot allocate memory' (errno=12)
> #
> # There is insufficient memory for the Java Runtime Environment to continue.
> # Native memory allocation (malloc) failed to allocate 320864256 bytes for committing reserved memory.
> # An error report file with more information is saved as:
> # /home/jenkins/agent-workspace/workspace/Plugins/git-plugin/3.0.0-beta@4/hs_err_pid12574.log
> Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000007e8c80000, 319815680, 0) failed; error='Cannot allocate memory' (errno=12)
> #
> # There is insufficient memory for the Java Runtime Environment to continue.
> # Native memory allocation (malloc) failed to allocate 319815680 bytes for committing reserved memory.
> # An error report file with more information is saved as:
> # /home/jenkins/agent-workspace/workspace/Plugins/git-plugin/3.0.0-beta@5/hs_err_pid12729.log
> Tests run: 696, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13.741 sec - in hudson.plugins.git.GitChangeSetPluginHistoryTest
> Running hudson.plugins.git.GitStatusTest
> Killed
>
> There's at least three things we can try:
> a. Add memory (or more swap space) to VM's
> b. Switch to jdk8 or investigate java settings
> c. Analyze tests to see what is pulling such a large chunk of memory.
>
> I think (a) is the most reasonable place to start. It makes sense that a
> machine would run short on memory when it runs multiple processes, and it
> is relatively easy to try.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/git-plugin/pull/411#issuecomment-230874648>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAJkDXn-8k67qqf-Q-TFWW5NsUng3dZLks5qS_41gaJpZM4Iyf6Y>
> .
>

bitwiseman commented:
I ran this through the VisualVM profiler in JDK7 and JDK8, for CPU and memory.    I'm not sure these help much.   Any other suggestions welcome.

JDK7:
<img width="1525" alt="screen shot 2016-07-06 at 3 00 50 pm" src="https://cloud.githubusercontent.com/assets/1958953/16636970/60b431f8-4390-11e6-9716-968c08e6e2d2.png">

JDK8:
<img width="1255" alt="screen shot 2016-07-06 at 3 24 49 pm" src="https://cloud.githubusercontent.com/assets/1958953/16636977/6abf2f36-4390-11e6-81e1-c7342de2547a.png">

bitwiseman commented:
@MarkEWaite - I've rebased this on master.  @rtyler added a "highmem" label to the jenkins-infra VMs with 8+gb RAM and I've added that label to the parallel test execution node specification.  

If you could push this to a branch in this repo, the infrastructure will pick it up and test it automatically - it doesn't need to be the 3.0.0-beta branch.  

Alternately, if you approve access for me to this repo, I could do this myself.  You can trust me to play nice.  😄

MarkEWaite commented:
Is there a reason it shouldn't just be merged into the 3.0.0-beta branch?  That will cause it to be re-evaluated each time a change is merged from master to 3.0.0-beta, and avoids having to keep another long-lived branch.

bitwiseman commented:
No, I just thought if it was any bother to you we could keep it out of your way.  Putting it in 3.0.0-beta is fine with me.  😄

bitwiseman commented:
@MarkEWaite - Please merge once more.  I removed my timeout steps because of the one you added in master, but then you reverted the one from master as well.  😄 I just added my timeout commit back here.

bitwiseman commented:
Arg!   Still failing with OOM. 
https://ci.jenkins.io/job/Plugins/job/git-plugin/job/3.0.0-beta/20

bitwiseman commented:
@MarkEWaite @rtyler - trying explicit memory settings.  This is setting it low, I'll try `1024m` if this  encounters issues.

MarkEWaite commented on Jenkinsfile:
```diff
@@ -81,7 +81,7 @@ void runParallelTests() {
 void mvn(def args) {
   /* Get jdk tool. */
-  String jdktool = tool name: "jdk7", type: 'hudson.model.JDK'
+  String jdktool = tool name: "jdk8", type: 'hudson.model.JDK'
```
Are you sure you want to switch to jdk8 for maven run?

The git client plugin must be compiled with jdk7.  The git plugin is compiled with jdk7 intentionally to reduce the risk of compatibility problems due to compiling with jdk8 but running on jdk7.

bitwiseman replied:
I was just trying different ideas.   JDK8 didn't help anyway. 😄

MarkEWaite commented on pom.xml:
```diff
@@ -53,6 +53,13 @@
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <argLine>-Xmx512m</argLine>
```
I tried a different technique (requires maven 3.3.1 or later). Didn't seem to resolve the issue for me.  

Refer to https://github.com/MarkEWaite/git-plugin/commit/8ad079cbbf636c98f311f3ce5437845d0523a0cb for the .mvn/jvm.config file technique that I used.  

The technique is described in the [maven 3.3.1 release notes](https://maven.apache.org/docs/3.3.1/release-notes.html#JVM_and_Command_Line_Options)

bitwiseman commented:
@MarkEWaite - I have set the memory limits even lower and made sure they apply to maven and surefire processes.  I've also reduced the parallelism to three nodes.  When you have chance to merge this to 3.0.0-beta, that'd be great.

MarkEWaite commented on pom.xml:
```diff
@@ -154,6 +161,11 @@
       <version>${workflow.version}</version>
     </dependency>
     <dependency>
+      <groupId>org.jenkins-ci.plugins.workflow</groupId>
```
Why did you add a dependency on the workflow-step-api plugin?  It already has that dependency for tests, and that is enough to compile and test the code.

I'd prefer you remove this before the change is merged to the 3.0.0-beta branch.

bitwiseman replied:
Hm, sorry.  Not sure why this is here. I'll remove it.

bitwiseman commented:
@MarkEWaite - Updated as you requested.

MarkEWaite commented:
My mistake, I used the "merge" button to merge the changes to the master branch.  That was unintentional.  I've used a "force push" to remove the changes from the master branch.

I intended to bring the changes into the 3.0.0-beta branch and will do that now.

MarkEWaite commented:
Merged to the 3.0 branch.

bitwiseman commented:
@MarkEWaite - Thanks!  It looks like this is finally stable in the infra test environment: 
https://ci.jenkins.io/job/Plugins/job/git-plugin/job/3.0.0-beta/

@rtyler - ran jobs 28 to 31 overlapping, and ran jobs 32 to 37 all at the same time.  It looks like the issue is the sheer number of JVM's using the default amount of memory, and restricting the machines down makes things work. 

Does this fail in any of your test VMs now?

MarkEWaite commented:
No obvious failures yet, but I need more run time before I am confident enough in the change to bring it to the master branch.

I had to make the PermGen change on the master branch because it was unreliable enough to be a distraction without that change.

bitwiseman commented:
Not bring into master yet: absolutely.  I'm perfectly happy to have this bake in 3.0-beta until it is proven.

Do you want to reopen this until then or are you okay with having it closed and in 3.0.0-beta?

MarkEWaite commented:
I'm OK with it in 3.0.0-beta. I plan to release a new 3.0.0-beta version soon (to match with git plugin 2.5.2), so I'll run some tests on the 3.0.0-beta branch to assure the changes are stable and well behaved.

MarkEWaite commented:
If you find there are additional changes needed, open a new pull request against the 3.0.0-beta branch (so that I can use the "Merge" button)

bitwiseman commented:
Sounds great!  Thanks so much for your patience and help!
